client/extensions/hello-dolly/state/reducer.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import { HELLO_DOLLY_NEXT_LYRIC } from './action-types';
+import {
+	ROUTE_SET,
+	SECTION_SET,
+	SITE_SETTINGS_SAVE,
+} from 'state/action-types.js';
+
+export default function lyricIndex( state = 0, action ) {
+	switch ( action.type ) {
+		case ROUTE_SET:
+		case SECTION_SET:
+		case SITE_SETTINGS_SAVE:
+		case HELLO_DOLLY_NEXT_LYRIC:
+			return advanceToNextLyric( state );
+		default:
+			return state;
+	}
+}
+
+function advanceToNextLyric( index ) {
+	return index + 1;
```

Comment:
I'd be ok with `hello-dolly` not adding anything to state for now as well and only using existing state through selectors to keep it simpler.

Not a strong opinion here, but it may be wiser to do the lyrics example once we figure out how to dynamically load this stuff.